sale_coupon_auto_refresh/models/sale_order.py

```diff
@@ -10,21 +12,17 @@ class SaleOrder(models.Model):
     auto_refresh_coupon = fields.Boolean(related="company_id.auto_refresh_coupon",)
 
     def _auto_refresh_coupons(self):
-        self.filtered(
-            lambda x: x.state in ("draft", "sent") and x.auto_refresh_coupon
-        ).recompute_coupon_lines()
-
-    @api.model_create_multi
-    def create(self, vals_list):
-        """Create or refresh coupon lines on create"""
-        orders = super().create(vals_list)
-        orders.with_context(skip_auto_refresh_coupons=True)._auto_refresh_coupons()
-        return orders
-
-    def write(self, vals):
-        """Create or refresh coupon lines after saving"""
-        res = super().write(vals)
-        if self.env.context.get("skip_auto_refresh_coupons"):
-            return res
-        self.with_context(skip_auto_refresh_coupons=True)._auto_refresh_coupons()
-        return res
+        func = type(self)._allow_recompute_coupon_lines
+        self.filtered(func).recompute_coupon_lines()
+
+    def _allow_recompute_coupon_lines(self) -> bool:
```

Comment:
I don't have a strong opinion on type hints in Odoo/OCA code, although they're rarely used...